main.cpp

```diff
@@ -8,49 +8,60 @@
 #endif
 
 wgpu::Instance instance;
+wgpu::Adapter adapter;
 wgpu::Device device;
 wgpu::RenderPipeline pipeline;
 
-wgpu::SwapChain swapChain;
+wgpu::Surface surface;
 const uint32_t kWidth = 512;
 const uint32_t kHeight = 512;
 
-void SetupSwapChain(wgpu::Surface surface) {
-  wgpu::SwapChainDescriptor scDesc{
-      .usage = wgpu::TextureUsage::RenderAttachment,
-      .format = wgpu::TextureFormat::BGRA8Unorm,
-      .width = kWidth,
-      .height = kHeight,
-      .presentMode = wgpu::PresentMode::Fifo};
-  swapChain = device.CreateSwapChain(surface, &scDesc);
+void ConfigureSurface() {
+  wgpu::SurfaceCapabilities capabilities;
+  surface.GetCapabilities(adapter, &capabilities);
+
+  wgpu::SurfaceConfiguration config = {
+    .device = device,
+    .width = kWidth,
+    .height = kHeight,
+    .format = capabilities.formats[0],
+    .alphaMode = capabilities.alphaModes[0],
+    .presentMode = capabilities.presentModes[0],
+  };
+
+  surface.Configure(&config);
 }
 
-void GetDevice(void (*callback)(wgpu::Device)) {
+void GetAdapter(void (*callback)(wgpu::Adapter)) {
   instance.RequestAdapter(
       nullptr,
       // TODO(https://bugs.chromium.org/p/dawn/issues/detail?id=1892): Use
-      // wgpu::RequestAdapterStatus, wgpu::Adapter, and wgpu::Device.
+      // wgpu::RequestAdapterStatus and wgpu::Adapter.
       [](WGPURequestAdapterStatus status, WGPUAdapter cAdapter,
          const char* message, void* userdata) {
         if (status != WGPURequestAdapterStatus_Success) {
           exit(0);
         }
         wgpu::Adapter adapter = wgpu::Adapter::Acquire(cAdapter);
-        adapter.RequestDevice(
-            nullptr,
-            [](WGPURequestDeviceStatus status, WGPUDevice cDevice,
-               const char* message, void* userdata) {
-              wgpu::Device device = wgpu::Device::Acquire(cDevice);
-              device.SetUncapturedErrorCallback(
-                  [](WGPUErrorType type, const char* message, void* userdata) {
-                    std::cout << "Error: " << type << " - message: " << message;
-                  },
-                  nullptr);
-              reinterpret_cast<void (*)(wgpu::Device)>(userdata)(device);
+        reinterpret_cast<void (*)(wgpu::Adapter)>(userdata)(adapter);
+  }, reinterpret_cast<void*>(callback));
+}
+
+void GetDevice(void (*callback)(wgpu::Device)) {
+  adapter.RequestDevice(
+      nullptr,
+      // TODO(https://bugs.chromium.org/p/dawn/issues/detail?id=1892): Use
+      // wgpu::Device.
+      [](WGPURequestDeviceStatus status, WGPUDevice cDevice,
+          const char* message, void* userdata) {
+        wgpu::Device device = wgpu::Device::Acquire(cDevice);
+        device.SetUncapturedErrorCallback(
```

Comment:
Note that this will be part of the descriptor soon.

Reply:
(It is already there in Dawn, but Emscripten doesn't implement it.)